### DISCUSSION TITLE
Which package do you usually put Mapper in, it is easy to conflict with the Mapper of mybatis, and will be scanned by mybatis

### DISCUSSION THREAD

**Discussion by liunancun** [mapstruct/mapstruct#2911] [General]
Which package do you usually put Mapper in, it is easy to conflict with the Mapper of mybatis, and will be scanned by mybatis

**Comment by Zegveld** [comment]
You can name your MapStruct mapper classes whatever you want. For example you could use the term `Converter` instead of `Mapper`, that way you don't get naming conflicts.

For example
```
@Mapper
interface ABConverter { // mapstruct will generate ABConverterImpl class.
    ObjectB convertAToB(ObjectA a);
}
```

The location where to place them depends on the usage. When mapping/converting from dto to domain I usually place them somewhere near the dto. Usually in a separate package. Unless for access reasons I need to have package access to a resource, then in the same package as the resource that enforces this restriction.

**Comment by Zegveld** [comment]
Also in gitter a same kind of question got asked:

kliarist
@kliarist
[20:15](https://gitter.im/mapstruct/mapstruct-users?at=62bb45333bbb384889bfa228)
I am guessing this question has come up before. Where should I be injecting my mappers? In the presentation layer (controllers) or application layer (services)? What is the best practice when it comes to DTOs? Shall they only be present on controllers or be used as input/output for methods in the services
Raimund Klein
@Chessray
[21:08](https://gitter.im/mapstruct/mapstruct-users?at=62bb51c81227f62be3e2b539)
That depends entirely on your general architecture approach. My personal favourite is proper DDD combined with a Ports and Adapters architecture. In that scenario, I've got mappers between REST layer DTOs and domain objects in the Controllers and mappers between domain objects and adapter value objects in the respective adapter modules.

Also converting this to a discussion since it is not an issue.

**Reply by kliarist** [reply]
I think I got the answer, which is "it depends". I kinda prefer having them injected in my services as I feel mapper logic is still business logic, especially when some mappers tend to be a lil more complex than usual. Also, when using jpa and relations, mapping to dtos should be performed while on a transactional context (otherwise dynamic proxies would moan about entity manager).. which to me means in the service classes..I wouldn't consider annotating my controllers with @Transactional. Thanks for the response
